frontend/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -0,0 +1,375 @@
+  columns = columns.map(column => {
+    if (column.id && columnSizes.has(column.id)) {
+      return {
+        ...column,
+        width: columnSizes.get(column.id),
+      } as GridColumnWithCellTemplate
+    }
+    return column
+  })
+
+  // Number of rows of the table minus 1 for the header row:
+  const numRows = element.isEmpty() ? 1 : element.dimensions.rows - 1
+  const numIndices = element.types?.index?.length ?? 0
+
+  const onColumnResize = React.useCallback(
+    (column: GridColumn, newSize: number) => {
+      if (column.id) {
+        setColumnSizes(new Map(columnSizes).set(column.id, newSize))
+      }
+    },
+    [columns]
+  )
+
+  const getCellContent = React.useCallback(
+    ([col, row]: readonly [number, number]): GridCell => {
+      if (element.isEmpty()) {
+        return {
+          ...getCellTemplate(ColumnType.Text, true, "faded"),
+          displayData: "empty",
+        } as GridCell
+      }
+
+      if (col > columns.length - 1) {
+        // This should never happen
+        return getCellTemplate(ColumnType.Text, true)
+      }
+
+      const cellTemplate = columns[col].getTemplate()
+      if (row > numRows - 1) {
+        // This should never happen
+        return cellTemplate
+      }
+      try {
+        // Quiver has the header in first row
+        const quiverCell = element.getCell(row + 1, col)
+        return fillCellTemplate(cellTemplate, quiverCell, element.cssStyles)
+      } catch (error) {
+        // This should not happen in read-only table.
+        logError(error)
+        return cellTemplate
+      }
+    },
+    [columns, numRows, element]
+  )
+
+  const { getCellContent: getCellContentSorted } = useColumnSort({
+    columns,
+    getCellContent,
+    rows: numRows,
+    sort,
+  })
+
+  if (sort !== undefined) {
+    columns = columns.map(column => {
+      if (column.id === sort.column.id) {
+        return {
+          ...column,
+          title:
+            sort.direction === "asc"
+              ? `↑ ${column.title}`
+              : `↓ ${column.title}`,
+        }
+      }
+      return column
+    })
+  }
+
+  return {
+    numRows,
+    numIndices,
+    columns,
+    getCellContent: getCellContentSorted,
+    onColumnResize,
+  }
+}
+export interface DataFrameProps {
+  element: Quiver
+  height?: number
+  width: number
+}
+
+function DataFrame({
+  element,
+  height: propHeight,
+  width: propWidth,
+}: DataFrameProps): ReactElement {
+  const [sort, setSort] = React.useState<ColumnSortConfig>()
+
+  const {
+    numRows,
+    numIndices,
+    columns,
+    getCellContent,
+    onColumnResize,
+  } = useDataLoader(element, sort)
+
+  const [isFocused, setIsFocused] = React.useState<boolean>(true)
+
+  const [gridSelection, setGridSelection] = React.useState<GridSelection>({
+    columns: CompactSelection.empty(),
+    rows: CompactSelection.empty(),
+  })
+
+  const dataEditorRef = React.useRef<DataEditorRef>(null)
+
+  const minWidth = MIN_COLUMN_WIDTH + 3
```

Comment:
Why do we have a MIN_COLUMN_WIDTH + 3 when we can just set MIN_COLUMN_WIDTH to 38?

Reply:
`MIN_COLUMN_WIDTH` and `minWidth` here are referring to different elements. The `minWidth` defines the min width of the table container, which is set to minimum one column + the borders + 1. On the other hand, `MIN_COLUMN_WIDTH`  is actually used as minimum width of any column.

Reply:
I will add a comment here to make that more clear.

Reply:
I moved it into a constant 👍  Since we need to have access to the `MIN_COLUMN_WIDTH` value anyways for other parts (e.g. for the calculation of the min table width, it's better to also set this value in theme as well. Otherwise, we would need to keep this value and the glide-data-grid default in sync.